Skill system framework #3918y7b

veteran29 opened the pull request:
### Acceptance Criteria
- [x] A skill can be defined declaratively in a config or SQF file
- [x] A skill can be unlocked by purchasing one or more other skills (using AND or OR boolean logic) - Callback functions for skills unlocking
- [x] A skill can be manually unlocked by a script. (e.g, when a player completes a certain action or visits a certain location). This is persistent.
- [x] An entire skill tree can be unlocked by one or more skills (e.g, specialization and expertise trees)
- [x] A skill tree has a parent tree, which will be used in the UI to give it a hierarchy.
- [x] A skill tree has a name, and a description, which will be used in the UI.
- [x] A skill can be purchased by spending a skill point, giving the player its benefit.
- [x] A passive skill is made active as soon as it is purchased, and persists across missions and respawn.
- [x] An active skill is triggerable by the player. 
- [x] An active skill has a cooldown period after activation in which it cannot be used.
- [x] The cooldown period affects each type of skill individually.
    - [x] E.g - Activating the primary skill doesn't start the ultimate skill cooldown.
    - [x] E.g - Activating the primary skill then swapping it out, means the new skill is still on cooldown.
- [x] One primary skill may be usable at a time. 
- [x] One ultimate skill may be usable at a time.
- ~~The ultimate skill must be lockable via script to prevent it being changed (e.g, when on a mission)~~
    - Should be handled in UI, IMO - @veteran29
- [x] Skills may have a criteria, which if not met means they will be invisible in the UI. 
- [x] Skills have a description, which will be shown in the UI.
- [x] Skills have a short description, which will be shown in the UI.
- [x] Skills have a name, which is shown in the UI.
- [x] Skills have an icon, which is shown in the UI.
- [x] Skills must be associated with a skill tree

veteran29 committed:
Skill tree config proto

veteran29 committed:
Update skill tree config

It's more in line with design docs now

veteran29 committed:
Add basic UI proto

Still a lot of to do

veteran29 committed:
Parse skills to hash

veteran29 committed:
Add default icon to skill template

veteran29 committed:
Handle skill tree drawing

veteran29 committed:
Add vgm_g_fnc_execNextFrame

veteran29 committed:
Allow traversing the tree both ways

veteran29 committed:
Make tree render unscheduled

veteran29 committed:
Add logging macros

veteran29 committed:
Revert "Add logging macros"

This reverts commit 0c7601dbfd354cbb6c22b24f94c650adf392b6aa.

veteran29 committed:
Add skill "path" to the hash

veteran29 committed:
Add tier and cost to skill hash

veteran29 committed:
Refactor skill tree parsing into dedicated function, move to global var

veteran29 committed:
Decrease amount of skill points when skill is learnt

veteran29 committed:
Add simple skill learning UI

veteran29 committed:
Fix skills_canLearn known skill check

veteran29 committed:
Fix serialization warnings in skill ui drawing

veteran29 committed:
Add skills_isKnown function

veteran29 committed:
Apply skill effect code upon learning

veteran29 committed:
Add green background to known skills

veteran29 committed:
Rudimentary server side skill learn handling

veteran29 committed:
Harden fnc_skills_learn input handling

veteran29 committed:
Prevent duplicated skill paths in DB data

veteran29 committed:
Rename subtrees to subtreesHash

veteran29 committed:
Add vgm_skills_pathsHash for faster path access

veteran29 committed:
Merge branch 'development' into skill-system

veteran29 committed:
Refactor skill learning handling into event based system

veteran29 committed:
Enable event system in core component

veteran29 committed:
Refactor skill handling to be server sided

veteran29 committed:
Add server-side skill learning validation

veteran29 committed:
Fix path of subtrees

veteran29 committed:
Check if previous tree was unlocked fully when in subtreee

SavageGameDesign commented:
Task linked: [CU-3918y7b Skills framework](https://app.clickup.com/t/3918y7b)

veteran29 committed:
Add vgm_g_fnc_skills_canSee

veteran29 committed:
Refactor handling of already known skills upon network sync

veteran29 committed:
Fix docs

veteran29 committed:
Execute global functions before client ones

veteran29 committed:
Add support for applyOnRespawn

veteran29 committed:
Update test skills config

veteran29 committed:
Fix receiveSkillsData not working properly

Typo in variable name

veteran29 committed:
Add active skills UI dev prototype

UI for testing and developing needed API functions

veteran29 committed:
Add skillWheel

veteran29 committed:
Add skill activation and cooldowns

veteran29 committed:
Fix cooldown check

veteran29 committed:
Disable ability2 skill slot

veteran29 committed:
Move fnc_isKnown to global functions

veteran29 committed:
Do not render hidden skills

veteran29 committed:
Add function for teching skills from server

Spoffy commented on game/configs/mission/cfg_skills.hpp:
```diff
@@ -0,0 +1,120 @@
+    conditionUnlock = "true";
+    conditionShow = "true";
+    codeApply = "systemChat 'skill applied'";
+    codeActivate = "systemChat 'skill activated'";
```
I think we'll need a hook for deactivating too, in case someone changes their skills

Spoffy replied:
We'll also need a cooldown property, for the delay between activations

Spoffy commented on game/functions/functions_client.hpp:
```diff
@@ -3,6 +3,41 @@
+        };
+    };
+
+    class skills
+    {
+        VGM_GLOBAL_PATH(\systems\skills\global);
+
+        class skills_canLearn {};
+        class skills_canSee {};
+        class skills_getByPath {};
+        class skills_isKnown {};
+        class skills_parseTreeCfg {};
+        class skills_preInit
+        {
+            preInit = 1;
+        };
+        class skills_treesHashToPathsHash;
```
I like your functions names, feels like I can intuit what these do easily enough.

Spoffy commented on game/functions/functions_client.hpp:
```diff
@@ -28,23 +63,48 @@ class vgm_c
+        class skills_active_preInit
         {
             preInit = 1;
         };
```
Can we call this preinit from the main skills one? It feels like active skills shouldn't be initialising separately from the rest of the skills system?

I think it makes it confusing for someone to deduce how the init process works. If they're reading the main init file to understand the system, they need to:
1.  Realise there's two init files
2.  Realise they're both configured in the functions config
3.  Understand that the load order is defined by the order in this file, which I don't think is common knowledge.

Spoffy commented on game/functions/functions_server.hpp:
```diff
@@ -25,4 +25,29 @@ class vgm_s
+
+        class skills_handle_skillsDataRequest {};
+        class skills_handle_skillLearnRequest {};
+    };
```
Nicely organised, I like your folder structure for the skills system.

Spoffy commented on game/functions/systems/skills/server/fn_skills_teachSkill.sqf:
```diff
@@ -0,0 +1,30 @@
+
+    Parameter(s):
+        _player - Player to give the skill point to [OBJECT]
+        _skillPath - Path to the skill being teched [ARRAY]
```
I think it should be `taught` here instead of `teached` :)

Spoffy commented on game/functions/systems/skills/server/fn_skills_dbGet.sqf:
```diff
@@ -0,0 +1,32 @@
+
+params ["_player"];
+
+private _hash = _player getVariable "vgm_g_skillsData";
```
By _hash, do you mean a skills hash table? As this confused me for a minute, I thought we were passing around some kind of `hashValue _skills` or similar :)

veteran29 replied:
It's a hashMap

Spoffy commented on game/functions/systems/skills/server/fn_skills_dbGet.sqf:
```diff
@@ -0,0 +1,32 @@
+if (isNil "_hash") then {
+    _hash = ["player_skills", getPlayerUID _player] call vgm_s_fnc_db_get;
+    _player setVariable ["vgm_g_skillsData", _hash];
+};
```
I think we should load player skills from the DB separately, rather than load it here.

My thinking is:
* If we move to an actual DB in the future, this would potentially be a blocking function (or we'd need to rewrite the db layer to have async options). Which would mean this whole function has a fast path, and a slow path, and we wouldn't know which we'd hit when calling it.
* It pulls from a cache, which I wouldn't have expected based on the name - I'd have expected a db fetch each time.

veteran29 replied:
If not load it here then where?

The slow path is done only once and it will always happen when a player is joining the server.

The idea behind the cache is that the player can't play on multiple servers at once, so we can re-use the data without polling the DB all the time.

Spoffy replied:
I'd load it on player join, before anything else happens, so that data is all there and waiting to be used.

In terms of cache - my point was more that I think the behaviour doesn't match the function name. I agree we should be caching it and not re-accessing the DB 🙂

gnalvesteffer replied:
I obviously don't have all the context here, but I do agree with @Spoffy regarding loading all the data upfront from the DB when a player joins, and then caching it, where feasible. Potentially long-running, blocking operations would introduce hitching during play, which is undesirable.

gnalvesteffer replied:
Also with that in mind ^, this function could be renamed to `cacheGet` or similar.

veteran29 replied:
> I'd load it on player join, before anything else happens, so that data is all there and waiting to be used.

Doesn't change much as first thing player does is call for this data == load it.

> I do agree with @Spoffy regarding loading all the data upfront from the DB when a player joins, and then caching it, where feasible. Potentially long-running, blocking operations would introduce hitching during play, which is undesirable.

It doesn't matter when we load the data if it's synchronous, loading it early during player connection will still freeze the server, unless we want to completely refactor the DB handling and make it async.


Anyways, is this good?: 0f7fa47d821f937268c97f9065482bfe15cefe9a

gnalvesteffer replied:
It should be async then, but it may or may not be such a big deal, depending on how frequently players join and the max player count. To your point, it wouldn’t matter that much if we load on join or whenever needed, since the server would freeze all the same to the clients. Though loading all data upfront on join would reduce the amount of freezing throughout a playthrough (given we need to pull more data).

Spoffy commented on game/functions/systems/skills/server/fn_skills_dbGet.sqf:
```diff
@@ -0,0 +1,32 @@
+};
+
+_hash getOrDefault ["skillPoints" , 0, true];
+_hash getOrDefault ["skillPaths" , [], true];
```
What do you think to using `_hash set ["skillPoints", 0, true]`, to be a bit a little clearer about what this operation is doing, since we're not using the value it returns?

Spoffy commented on game/functions/systems/skills/server/fn_skills_dbGet.sqf:
```diff
@@ -0,0 +1,32 @@
+private _hash = _player getVariable "vgm_g_skillsData";
+if (isNil "_hash") then {
+    _hash = ["player_skills", getPlayerUID _player] call vgm_s_fnc_db_get;
+    _player setVariable ["vgm_g_skillsData", _hash];
```
Will there be a runtime error if there's no skills data in the database? Specifically when we hit the `getOrDefault` lines below

veteran29 replied:
`db_get` always returns a HashMap.

Spoffy replied:
Cool - you're absolutely right.

Spoffy commented on game/functions/systems/skills/server/fn_skills_dbSave.sqf:
```diff
@@ -0,0 +1,31 @@
+if (isNil "_hash") exitWith {
+    ["ERROR", format ["VGM: No data for player %1 (%2)", name _player, _uid]] call para_g_fnc_log;
+    false // return
+};
```
Where is the skills data structure initially set on new players? As I can't see dbGet or dbSet handling it, and I can't find the init code elsewhere, but I very well could just be missing it.

veteran29 replied:
`dbGet` L26

Spoffy replied:
Aha, got it, it's a side-effect of the dbGet, I see that now. Thanks!

Spoffy commented on game/functions/systems/skills/server/fn_skills_dbSave.sqf:
```diff
@@ -0,0 +1,31 @@
+};
+
+["player_skills", _uid, _hash] call vgm_s_fnc_db_typed_save;
+saveMissionProfileNamespace // return
```
I think this shouldn't a concern of the skills system, feels like this should be transparently handled by the db system, as we shouldn't need to care about the backend saving implementation.

Spoffy commented on game/functions/systems/skills/server/fn_skills_dbSave.sqf:
```diff
@@ -0,0 +1,31 @@
+params ["_player"];
+
+private _uid = getPlayerUID _player;
+private _hash = _player call vgm_s_fnc_skills_dbGet;
```
Similar comment to above about naming, I wonder if we should use `_skillsHashMap` or `_playerSkills` or similar?

Spoffy commented on game/functions/systems/skills/server/fn_skills_teachSkill.sqf:
```diff
@@ -0,0 +1,30 @@
+/*
+    File: fn_skills_teachSkill.sqf
```
I can't find any calls to this function - is it just a utility function for debugging?

veteran29 replied:
>A skill can be manually unlocked by a script. (e.g, when a player completes a certain action or visits a certain location). This is persistent.

API Func. Part of acceptance criteria.

Spoffy commented on game/functions/systems/skills/client/network/fn_skills_receiveSkillsData.sqf:
```diff
@@ -0,0 +1,48 @@
+        continue;
+    };
+
+    if (_skill getOrDefault ["_known", false]) then {
```
Why do we use `_` at the start of this hashmap property here?

veteran29 replied:
The skillTrees hashmap is identical on every client by default and is kinda "locally global", properties that are not coming from config but are used locally are marked with `_`.

Spoffy replied:
Ah, got it. I think we need to document that somewhere, maybe in README.md in the skills system directory?

veteran29 replied:
Not relevant anymore, the list of already applied skills will be kept in an array.

Spoffy commented on game/functions/systems/skills/client/network/fn_skills_receiveSkillsData.sqf:
```diff
@@ -0,0 +1,48 @@
+        continue;
+    };
+
+    _skill set ["_known", true];
```
When the player receives skill data, I don't think we roll back their old skills data? So for example, if we un-taught a skill on the server, it would always be set as known on the client.

veteran29 replied:
Added handling of forgotten skills with skill respec.

Spoffy commented on game/functions/systems/skills/client/network/fn_skills_receiveSkillsData.sqf:
```diff
@@ -0,0 +1,48 @@
+
+    ["DEBUG", format ["VGM: New skill '%1'", _x]] call para_g_fnc_log;
+
+    ["vgm_skills_learnt", [_x, _skill]] call para_g_fnc_event_trigger;
```
If this is called whenever we learn a skill, for every skill, I think that means the `apply` code for each skill would need to be idempotent? 

For example, it couldn't be `player setCustomAimCoef (getCustomAimCoef player + 1)`, every time a player learned a different skill it'd give them the bonus for that skill again.

I think this poses some interesting challenges, such as multiple skills which modify the same thing on the player, but not unsolvable ones.

There's three options I think we could take here:
* Only apply newly learned skills
* Remove all skills, and re-apply them all.
* Add documentation to the skills config that `apply` needs to be idempotent, and explain the consequences.

Given idempotency can be quite a nice property (it'll reduce bugs most likely, although increasing complexity), I'd be keen to hear what you think.

veteran29 replied:
This is not a concern.

Following applies due to `_known` check:
> Only apply newly learned skills

Spoffy replied:
Right, I didn't see that check! Thanks :)

Spoffy commented on game/functions/systems/skills/server/network/fn_skills_handle_skillLearnRequest.sqf:
```diff
@@ -0,0 +1,44 @@
+        Something [BOOL]
+ */
+
+params ["_player", "_skillPath"];
```
Do we need to do `owner _player isEqualTo remoteExecutedOwner` for security here?

veteran29 replied:
Not sure if we need it but added.

Spoffy commented on game/functions/systems/skills/global/fn_skills_preInit.sqf:
```diff
@@ -0,0 +1,19 @@
+ */
+
+vgm_skills_treesHash = [missionConfigFile >> "vgm_skillTrees"] call vgm_g_fnc_skills_parseTreeCfg;
+vgm_skills_pathsHash = [vgm_skills_treesHash] call vgm_g_fnc_skills_treesHashToPathsHash;
```
I like that you've cached these in hashmaps, great idea. Wonder if we should add "map" to the end of "Hash" to make it clear we've got a map?

Spoffy commented on game/configs/mission/cfg_skills.hpp:
```diff
@@ -0,0 +1,120 @@
+    description = "SKILL DESCRIPTION";
+    icon = "\vn\ui_f_vietnam\ui\wheelmenu\img\ui_icon_a_ca.paa";
+
+    isActive = 0; // 1 - primary skill, 2 - ultimate skill
```
Can we use `skillType` or similar for this? Since this is prefixed with `is`, I think I'd expect it to be a boolean value

Spoffy commented on game/functions/systems/skills/global/fn_skills_parseTreeCfg.sqf:
```diff
@@ -0,0 +1,82 @@
+                ["conditionUnlock", compileFinal getText (_x >> "conditionUnlock")],
+                ["conditionShow", compileFinal getText (_x >> "conditionShow")],
+                ["codeApply", compileFinal getText (_x >> "codeApply")],
+                ["codeActivate", compileFinal getText (_x >> "codeActivate")]
```
Does compileFinal work in hashmaps? This isn't feedback, I'm just curious :)

veteran29 replied:
It does not, matter of habit.

Spoffy commented on game/functions/systems/skills/client/fn_skills_getSkillTreeFromSkill.sqf:
```diff
@@ -0,0 +1,32 @@
+
+private _item = vgm_skills_treesHash get (_path deleteAt 0);
+while {true} do {
+    if (count _path == 1) exitWith {
```
Can we do `<= 1` here? 

On the happy path where everything works, it'll behave the same.

On the unhappy path (where path ends up as 0, if we *somehow* have bad data in skill), it prevents an infinite loop.

Not exactly a likely issue, but it's a little bit more defensive :)

Can add a comment why it's `<=` for clarity too.

Spoffy commented on game/functions/systems/skills/client/fn_skills_getSkillPoints.sqf:
```diff
@@ -0,0 +1,21 @@
+/*
+    File: fn_skills_getSkillPoints.sqf
```
To be 100% clear, is this the skill points the player has left to spend, and not their total?

veteran29 replied:
Current amount of skill points that player can spend.

Spoffy commented on game/functions/systems/skills/client/active/fn_skills_active_openSkillWheel.sqf:
```diff
@@ -0,0 +1,44 @@
+    // ,"ability2"
+];
+
+[_iconsArray] call vn_fnc_wm_init;
```
We've got a wrapper around the native wheel menu here, that might be easier to use, and has some convenience functions (such as loading a wheel menu interface from config)
paradigm\client\functions\ui\wheel_menu\fn_wheel_menu_open.sqf

veteran29 replied:
Not a concern for this PR, it's a placeholder UI.

Spoffy commented on game/functions/systems/skills/client/active/fn_skills_active_preInit.sqf:
```diff
@@ -0,0 +1,54 @@
+    };
+
+    [] call vgm_c_fnc_skills_active_openSkillWheel;
+}];
```
Short term, all good with this.

Long term, we should integrate the keybindings menu from MF, or use Arma's builtin one. I'll open a dev ticket for it.

veteran29 replied:
Not a concern for this PR, it's a placeholder UI. Arma modded keybindings can only be used in Mods.

Spoffy commented on game/functions/systems/skills/client/active/fn_skills_active_preInit.sqf:
```diff
@@ -0,0 +1,54 @@
+
+    vgm_c_skills_active_list set [_path, _skill];
+
+}] call para_g_fnc_event_subscribeLocal;
```
We'll need to handle unlearning here too.

veteran29 replied:
Added handling of forgotten skills with skill respec.

Spoffy commented on game/functions/systems/skills/client/active/fn_skills_active_isSlotOnCooldown.sqf:
```diff
@@ -0,0 +1,29 @@
+    _slot = vgm_c_skills_active_slots getOrDefault [_slot, createHashMap];
+};
+
+time < (_slot getOrDefault ["cooldown", 1e10]) // return
```
I'd have expected cooldown to be "how long this takes to cooldown" from the name, such as 60 seconds, or 20 seconds remaining.

Maybe we could use "nextAvailableTime" or "cooldownEndTime" to be clearer :)

Spoffy commented on game/functions/systems/skills/client/active/fn_skills_active_skillWheelActivate.sqf:
```diff
@@ -0,0 +1,30 @@
+};
+
+player call (_skill get "codeActivate");
+private _cooldown = time + ([30, 60] select (_skill get "isUltimate"));
```
Same as the config comment, it'd be good to have this as a per-skill setting :)

Spoffy commented on game/functions/systems/skills/global/fn_skills_parseTreeCfg.sqf:
```diff
@@ -0,0 +1,82 @@
+            createHashMapFromArray [
+                ["path", _path + [configName _x]],
+                ["tier", _tier],
+                ["cost", 1],
```
Good thinking, can we pull this out into the config too? :)

Spoffy commented on game/functions/systems/skills/server/fn_skills_addSkillPoint.sqf:
```diff
@@ -0,0 +1,33 @@
+
+if (_broadcast) then {
+    [_skillsData] remoteExecCall ["vgm_c_fnc_skills_receiveSkillsData", _player];
+};
```
Wonder if we could/should add a skill point without re-broadcasting all the skills data? As I think this will trigger them all to be re-learned, etc.

Spoffy commented on game/functions/systems/skills/global/fn_skills_canLearn.sqf:
```diff
@@ -0,0 +1,61 @@
+        _path = _path select [0, count _path - 2];
+        private _skillTree = _path call vgm_g_fnc_skills_getByPath;
+        private _tiersCount = count (_skillTree get "skills");
+        diag_log [_skillTree get "displayName", count (_skillTree get "skills")];
```
Rogue diag_log. Maybe we put it as a debug log, or cut it? :)

Spoffy commented on game/functions/systems/skills/global/fn_skills_canLearn.sqf:
```diff
@@ -0,0 +1,61 @@
+        // no previous tree, always unlocked
+        if (count _path < 3) exitWith {true};
+
+        _path = _path select [0, count _path - 2];
```
Could we put a comment in here, explaining we're selecting the skill's tree's parent tree? :)

Spoffy commented on game/functions/systems/skills/global/fn_skills_canLearn.sqf:
```diff
@@ -0,0 +1,61 @@
+        if (count _path < 3) exitWith {true};
+
+        _path = _path select [0, count _path - 2];
+        private _skillTree = _path call vgm_g_fnc_skills_getByPath;
```
Same here, maybe we rename this to `parentSkillTree` or similar for clarity :)

Spoffy commented on game/functions/systems/skills/server/fn_skills_addSkillPoint.sqf:
```diff
@@ -0,0 +1,33 @@
+private _skillsData = _player call vgm_s_fnc_skills_dbGet;
+private _skillPoints = _skillsData get "skillPoints";
+
+_skillsData set ["skillPoints", _skillPoints + 1];
```
Is there a missing skills_dbSave here?

Spoffy commented on game/functions/systems/skills/server/network/fn_skills_handle_skillLearnRequest.sqf:
```diff
@@ -0,0 +1,44 @@
+
+// decrease amount of skill points
+private _skillPoints = _skillsData get "skillPoints";
+_skillsData set ["skillPoints", _skillPoints - 1];
```
this is currently -1, I think we should be grabbing the skill's cost?

veteran29 replied:
Skill cost wasn't fully implemented, fixed.

Spoffy commented on game/functions/systems/skills/server/fn_skills_addSkillPoint.sqf:
```diff
@@ -0,0 +1,33 @@
+        [player, true] call vgm_s_fnc_skills_addSkillPoint
+ */
+
+params ["_player", ["_broadcast", false]];
```
What's the situation where not sync'ing this with the client would help us?  I'm a little concerned the default settings here might lead to us accidentally de-syncing the skills from the client.

veteran29 replied:
The idea was that it can be executed multiple times before broadcasting but it does not make sense, I've added `_amount` param instead if you want to add multiple skill points.

Spoffy commented on game/functions/systems/skills/server/network/fn_skills_handle_skillLearnRequest.sqf:
```diff
@@ -0,0 +1,44 @@
+};
+
+
+["SUCCESS", format ["VGM: Handling learn request for %1 (%2), %3", name _player, getPlayeRUID _player, _skillPath]] call para_g_fnc_log;
```
We probably want to use "INFO" or "DEBUG", as I don't think "SUCCESS" is one of the standard (undocumented) log levels. I'll push to paradigm later with the header updated with a list of log levels.

Spoffy commented on game/functions/systems/skills/server/fn_skills_teachSkill.sqf:
```diff
@@ -0,0 +1,30 @@
+
+[_player, _skillsData] call vgm_s_fnc_skills_dbSave;
+
+[_skillsData] remoteExecCall ["vgm_c_fnc_skills_receiveSkillsData", _player];
```
Feels like this duplicates handle_skillLearnRequest a lot. Maybe we should have handle_skillLearnRequest call this? (Wondering if that was your original plan, since this is unused currently)

veteran29 replied:
Refactored it to use `teachSkill` in `skillLearnRequest`.

Spoffy reviewed (changes requested):
Phew, big PR! Left a bunch of feedback. I don't think there's major or system changing, just lots of little tweaks and a few bits of added functionality.

The biggest thing I think we need to add is unlearning skills. I forgot to spec it out in the original ticket I think ☹️

veteran29 committed:
Add configurable skill cooldown

veteran29 committed:
Init active skills from main skills preInit

veteran29 committed:
Enable configurable skill cooldown

veteran29 committed:
Fix typo

veteran29 committed:
Remove _broadcast and add _amount param to addSkillPoint

veteran29 committed:
Use teachSkill in learn requests

veteran29 committed:
Add skill costs to skill config

veteran29 committed:
Handle skill cost server side

veteran29 committed:
Use set instead of getOrDefault in dbGet

veteran29 committed:
Use _hashMap instead of _hash, do not save profileNamespace

gnalvesteffer commented on game/functions/systems/skills/server/fn_skills_dbGet.sqf:
```diff
@@ -0,0 +1,32 @@
+
+params ["_player"];
+
+private _hash = _player getVariable "vgm_g_skillsData";
```
Would recommend naming this `_skillsData`. `_hash` just tells me the datatype (or really it should be `_hashMap` if you want to be accurate). Alternatively `_skillsHashMap`.

gnalvesteffer commented on game/functions/systems/skills/server/fn_skills_dbGet.sqf:
```diff
@@ -0,0 +1,32 @@
+        _player - Player to get the skill data for [OBJECT]
+
+    Returns:
+        Something [BOOL]
```
Is this Return documentation accurate? Looks like it returns the skills data for the player (hashmap).

veteran29 committed:
Refactor player data fetching to run onPlayerConnected

veteran29 committed:
Add owner checks for request functions

veteran29 committed:
Rename global skill hashmap variables

veteran29 committed:
Prevent infinite loop in case of edgecase in getSkillTreeFromSkill

veteran29 committed:
Introduce vgm_c_skills_knownPathsList instead of _known var on skillHash

veteran29 committed:
Rename vgm_c_skills_knownPathsList to vgm_c_skills_appliedSkillsPaths

veteran29 committed:
Add codeUnapply to skill config

veteran29 committed:
Handle skills being forgotten

veteran29 committed:
Fix skills loading on Local hosted

veteran29 committed:
Add skillForget API func

Might get handy

veteran29 committed:
Clear forgotten skill from ability slots

veteran29 committed:
Track amount of spent skill points

veteran29 committed:
Implement skills respec

veteran29 committed:
Refactor forgerSkill into forgetSkills

veteran29 committed:
Apply review suggestions

veteran29 committed:
Rename CfgSkill isActive to skillType

veteran29 committed:
Add veteran29 to admin list

veteran29 committed:
Merge remote-tracking branch 'origin/development' into skill-system

veteran29 committed:
Refactor player skills data fetching to run on demand instead of EH

Event handler approach is problematic as players can be initialized before PlayerConnected fires

veteran29 committed:
Fix canLearn on dedicated server

veteran29 committed:
Remove diag_logs

Spoffy commented on game/functions/systems/skills/server/fn_skills_dbGet.sqf:
```diff
@@ -0,0 +1,32 @@
+/*
+    File: fn_skills_dbSave.sqf
```
Header needs an update :)

veteran29 replied:
```suggestion
    File: fn_skills_dbGet.sqf
```

veteran29 committed:
Update dbGet header
